torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -0,0 +1,109 @@
+import math
+
+import numpy as np
+import torch
+import torch.distributed as dist
+
+
+def _orthogonalize(matrix, epsilon=1e-8):
+    """Applies Gram-Schmidt procedure to orthogonalize a given 2D tensor."""
+    num_cols = matrix.shape[1]
+    for i in range(num_cols):
+        # Normalize the i'th column.
+        col = matrix[:, i : i + 1]
+        # Note that col ** 2 can underflow/overflow if we use FP16.
+        # May need to consder multiplying a scaling factor and divding it later, or using bfloat16 isntead.
+        col /= torch.sqrt(torch.sum(col ** 2))
+        # If no epsilon is added here, division by zero may be caused by vanishing gradients.
+        # This epsilon is not needed if the input matrix covers the gradients of at least one entire layer in the neural network.
+        if epsilon > 0:
+            col += epsilon
+        # Project it on the rest and remove it.
+        if i + 1 < num_cols:
+            rest = matrix[:, i + 1 :]
+            rest -= torch.sum(col * rest, dim=0) * col
+
+
+def _set_random(vector, random_seed=0):
+    """
+    Initializes a 1D tensor from a standard normal distribution.
+    The see makes sure that the randomized vector in all the DDP replicas are the same at every step.
+    """
+    torch.manual_seed(np.random.RandomState(random_seed).randint(1_000_000_000))
```

Comment:
If the np random number generator is only used once, I don't see the benefit of having it. You could just pass `random_seed` here directly. Maybe this shouldn't really be a separate function. You could just write something like:
```python
with torch.random.fork_rng():
    torch.manual_seed(0)  # if you do this multiple times, you should not repeatedly use the same seed
    vector.data[:] = torch.randn_like(vector)
```
in place of line 83 below.

Reply:
Thanks!